compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts

```diff
@@ -24,6 +35,341 @@ import {
+ *
+ * Example:
+ * <>foo</>
+ * _______________
+ * let t1;
+ * if (__DEV__) {
+ *   t1 = <>foo</>
+ * } else {
+ *   t1 = {...}
+ * }
+ *
+ */
+export function inlineJsxTransform(
+  fn: HIRFunction,
+  inlineJsxTransformConfig: NonNullable<
+    EnvironmentConfig['inlineJsxTransform']
+  >,
+): void {
+  const inlinedJsxDeclarations: InlinedJsxDeclarationMap = new Map();
+  /**
+   * Step 1: Codegen the conditional and ReactElement object literal
+   */
+  for (const [_, currentBlock] of [...fn.body.blocks]) {
+    let fallthroughBlockInstructions: Array<Instruction> | null = null;
+    const instructionCount = currentBlock.instructions.length;
+    for (let i = 0; i < instructionCount; i++) {
+      const instr = currentBlock.instructions[i]!;
+      switch (instr.value.kind) {
+        case 'JsxExpression':
+        case 'JsxFragment': {
+          /**
+           * Split into blocks for new IfTerminal:
+           *   current, then, else, fallthrough
+           */
+          const currentBlockInstructions = currentBlock.instructions.slice(
+            0,
+            i,
+          );
+          const thenBlockInstructions = currentBlock.instructions.slice(
+            i,
+            i + 1,
+          );
+          const elseBlockInstructions: Array<Instruction> = [];
+          fallthroughBlockInstructions ??= currentBlock.instructions.slice(
+            i + 1,
+          );
+
+          const fallthroughBlockId = fn.env.nextBlockId;
+          const fallthroughBlock = {
+            ...currentBlock,
+            id: fallthroughBlockId,
+            instructions: fallthroughBlockInstructions,
+          };
+
+          /**
+           * Complete current block
+           * - Add instruction for variable declaration
+           * - Add instruction for LoadGlobal used by conditional
+           * - End block with a new IfTerminal
+           */
+          const varPlace = createTemporaryPlace(fn.env, instr.value.loc);
+          promoteTemporary(varPlace.identifier);
+          const varLValuePlace = createTemporaryPlace(fn.env, instr.value.loc);
+          const thenVarPlace = {
+            ...varPlace,
+            identifier: forkTemporaryIdentifier(
+              fn.env.nextIdentifierId,
+              varPlace.identifier,
+            ),
+          };
+          const elseVarPlace = {
+            ...varPlace,
+            identifier: forkTemporaryIdentifier(
+              fn.env.nextIdentifierId,
+              varPlace.identifier,
+            ),
+          };
+          const varInstruction: Instruction = {
+            id: makeInstructionId(0),
+            lvalue: {...varLValuePlace},
+            value: {
+              kind: 'DeclareLocal',
+              lvalue: {place: {...varPlace}, kind: InstructionKind.Let},
+              type: null,
+              loc: instr.value.loc,
+            },
+            loc: instr.loc,
+          };
+          currentBlockInstructions.push(varInstruction);
```

Comment:
Hmm this isn't safe. There are two main categories of blocks which correspond roughly to statement-level blocks (kind 'block') and expression-level blocks (kind 'logical' etc). It's only safe to add a variable declaration at a block statement. So i think what you'd need to do is write this as a ternary, with an unnamed (non-promoted) variable that is getting assigned one of two different values without a variable declaration.

ie you're doing:

```js
let t0;
if (__DEV__) {
  t0 = { $$typeof: 'react.element', ...}
} else {
  t0 = jsx(...);
}
```

When it needs to be 

```js
__DEV__ ? { $$typeof: ... } : jsx(...)
```

So that it works in statement and expression contexts.

Ternaries are a bit tedious to create, you need both a ternary terminal and a branch terminal. But you get to avoid the variable declaration. You'll still need to create the phi in the fallthrough block.